core/embed/trezorhal/optiga/optiga_transport.c

```diff
@@ -0,0 +1,497 @@
+      }
+
+      ret = optiga_i2c_read(frame_buffer, size);
+      if (OPTIGA_SUCCESS != ret) {
+        return ret;
+      }
+
+      if (calc_crc(frame_buffer, size - 2) !=
+          (frame_buffer[size - 2] << 8) + frame_buffer[size - 1]) {
+        return OPTIGA_ERR_CRC;
+      }
+
+      frame_size = size - 2;
+
+      return OPTIGA_SUCCESS;
+    }
+  }
+
+  return OPTIGA_ERR_TIMEOUT;
+}
+
+static optiga_result optiga_send_packet(uint8_t packet_control_byte,
+                                        const uint8_t *packet_data,
+                                        size_t packet_data_size) {
+  if (packet_data_size > sizeof(frame_buffer) - 7) {
+    return OPTIGA_ERR_SIZE;
+  }
+
+  // Set up frame control information.
+  uint8_t fctr = FTYPE_DATA | SEQCTR_ACK;
+  fctr |= (frame_num_out << FRNR_SHIFT) & FRNR_MASK;
+  fctr |= (frame_num_in & ACKNR_MASK);
+
+  // Compile frame.
+  frame_buffer[0] = REG_DATA;
+  frame_buffer[1] = fctr;
+  frame_buffer[2] = (packet_data_size + 1) >> 8;
+  frame_buffer[3] = (packet_data_size + 1) & 0xff;
+  frame_buffer[4] = packet_control_byte;
+  memcpy(&frame_buffer[5], packet_data, packet_data_size);
+  uint16_t crc = calc_crc(&frame_buffer[1], packet_data_size + 4);
+  frame_buffer[packet_data_size + 5] = crc >> 8;
+  frame_buffer[packet_data_size + 6] = crc & 0xff;
+
+  return optiga_i2c_write(frame_buffer, packet_data_size + 7);
+}
+
+static optiga_result optiga_receive_packet(uint8_t *packet_control_byte,
+                                           uint8_t *packet_data,
+                                           size_t max_packet_data_size,
+                                           size_t *packet_data_size) {
+  // Expected frame control information byte.
+  uint8_t fctr = FTYPE_DATA | SEQCTR_ACK;
+  fctr |= (frame_num_in << FRNR_SHIFT) & FRNR_MASK;
+  fctr |= (frame_num_out & ACKNR_MASK);
+  *packet_data_size = (frame_buffer[1] << 8) + frame_buffer[2] - 1;
+  if (frame_size < 3 || frame_buffer[0] != fctr ||
+      *packet_data_size + 4 != frame_size) {
+    frame_size = 0;
+    return OPTIGA_ERR_UNEXPECTED;
+  }
+  frame_size = 0;
+
+  if (*packet_data_size > max_packet_data_size) {
+    return OPTIGA_ERR_SIZE;
+  }
+
+  *packet_control_byte = frame_buffer[3];
+  memcpy(packet_data, &frame_buffer[4], *packet_data_size);
+  return OPTIGA_SUCCESS;
+}
+
+optiga_result optiga_execute_command(
+    bool presentation_layer, const uint8_t *command_data, size_t command_size,
+    uint8_t *response_data, size_t max_response_size, size_t *response_size) {
+  *response_size = 0;
+  optiga_result ret = optiga_ensure_ready();
+  if (ret != OPTIGA_SUCCESS) {
+    return ret;
+  }
+
+  uint8_t pctr = 0;
+  if (presentation_layer) {
+    pctr |= PCTR_PRESENTATION_LAYER;
+  }
+
+  // Transmit command packets to OPTIGA.
+  uint8_t chain = PCTR_CHAIN_NONE;
+  do {
+    size_t packet_data_size = 0;
+    if (command_size > OPTIGA_MAX_PACKET_SIZE - 1) {
+      packet_data_size = OPTIGA_MAX_PACKET_SIZE - 1;
```

Comment:
The `-1` in `OPTIGA_MAX_PACKET_SIZE - 1` seems weird, like either the max packet size is off or there was one terminator byte expected.
The `optiga_send_packet` does not use `OPTIGA_MAX_PACKET_SIZE` for size check but computes it from `packet_data_size > sizeof(frame_buffer) - 7` which makes it hard to read if the checks are equivalent and correct.

Reply:
The first byte of the packet is the packet control byte `pctr`, so each packet contains at most `OPTIGA_MAX_PACKET_SIZE - 1` bytes of data. I can add a code comment to explain that or define another constant  `OPTIGA_MAX_PACKET_DATA_SIZE = OPTIGA_MAX_PACKET_SIZE - 1`. But since this is the only place where it would be used, I think the code comment would be better, so that the names don't get confused.

Reply:
The reason why `optiga_send_packet()` compares the `packet_data_size` with the size of the `frame_buffer` is because it will be writing the data to the `frame_buffer`. If someone were to accidentally change the `frame_buffer` size to something smaller, then this runtime check would catch it, but comparing against `OPTIGA_MAX_PACKET_SIZE` wouldn't catch it.

Reply:
Comment would be OK. Originally I thought that the frame size is given by the protocol and can't be changed.
We could make a compile-time assert (needs changing `gnu99` to `gnu11` in `SConscript.firmware`) to make sure the size relations are correct to avoid wrong change, e.g.:

```C
#include <assert.h>
// after macro and variable definitions:
static_assert(sizeof(frame_buffer) - 7 == OPTIGA_MAX_PACKET_SIZE-1,
        "incorrect frame size vs data max size constants relation" );
```

Reply:
In other parts of our codebase we use `_Static_assert()` and it seems to work fine.
https://github.com/trezor/trezor-firmware/pull/3186/commits/ac3e75470f77191a20e1b18a2694438cd708abaa https://github.com/trezor/trezor-firmware/pull/3186/commits/51ebf2b820ad99ffe4ed7c88541ff2c75776b3f8